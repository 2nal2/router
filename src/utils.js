export { pick, match, resolve };

function pick(routes, uri) {
  let match;
  let default_;

  let [uriPathname] = uri.split("?");
  let uriSegments = segmentize(uriPathname);
  let ranked = rankRoutes(routes);
  let matchedSegmentCount = 0;

  for (let i = 0, l = ranked.length; i < l; i++) {
    let route = ranked[i].route;

    if (route.default) {
      default_ = route;
      continue;
    }

    let routeSegments = segmentize(route.pattern);
    let params = {};

    for (let i = 0, l = routeSegments.length; i < l; i++) {
      let routeSegment = routeSegments[i];
      let uriSegment = uriSegments[i];

      if (uriSegment === undefined) {
        matchedSegmentCount = 0;
        break;
      }

      let isSplat = routeSegment === "*";
      if (isSplat) {
        params["*"] = uriSegments
          .slice(i)
          .map(decodeURIComponent)
          .join("/");
        matchedSegmentCount = i;
        break;
      }

      let dynamicMatch = paramRe.exec(routeSegment);
      if (dynamicMatch) {
        let value = decodeURIComponent(uriSegment);
        params[dynamicMatch[1]] = value;
      } else if (routeSegment !== uriSegment) {
        matchedSegmentCount = 0;
        break;
      }

      matchedSegmentCount++;
    }

    if (matchedSegmentCount) {
      match = { route, params };
      break;
    }
  }

  if (match) {
    return Object.assign({}, match, {
      uri: "/" + uriSegments.slice(0, matchedSegmentCount).join("/")
    });
  }

  if (default_) {
    return {
      route: default_,
      params: {},
      uri
    };
  }

  return null;
}

function match(pattern, uri) {
  return pick([{ pattern }], uri);
}

function resolve(to, base) {
  if (to.startsWith("/")) {
    // /foo/bar, /baz/qux => /foo/bar
    return to;
  }

  let [toPathname, toQuery] = to.split("?");
  let [basePathname, baseQuery] = base.split("?");

  let toSegments = segmentize(toPathname);
  let baseSegments = segmentize(basePathname);

  if (toSegments[0] === "") {
    // ?a=b, /users?b=c => /users?a=b
    return addQuery(basePathname, toQuery);
  }

  if (!toSegments[0].startsWith(".")) {
    // 456, /users/789 => /users/456
    let pathname =
      "/" +
      baseSegments
        .slice(0, baseSegments.length - 1)
        .concat(toSegments)
        .join("/");
    return addQuery(pathname, toQuery);
  }

  let segments = baseSegments
    .slice(0, baseSegments.length - 1)
    .concat(toSegments);

  let ups = 0;
  for (let i = segments.length - 1; i >= 0; i--) {
    const segment = segments[i];
    if (segment === ".") {
      segments.splice(i, 1);
    } else if (segment === "..") {
      segments.splice(i, 1);
      ups++;
    } else if (ups > 0) {
      segments.splice(i, 1);
      ups--;
    }
  }

  return addQuery("/" + segments.join("/"), toQuery);
}

////////////////////////////////////////////////////////////////////////

let paramRe = /^:(.+)/;

let SEGMENT_POINTS = 4;
let STATIC_POINTS = 3;
let DYNAMIC_POINTS = 2;
let SPLAT_POINTS = 1;
let ROOT_POINTS = 1;

let isRootSegment = segment => segment == "";
let isDynamic = segment => paramRe.test(segment);
let isSplat = segment => segment === "*";

let rankRoute = (route, index) => {
  let score = route.default
    ? 0
    : segmentize(route.pattern).reduce((score, segment) => {
        score += SEGMENT_POINTS;
        if (isRootSegment(segment)) score += ROOT_POINTS;
        else if (isDynamic(segment)) score += DYNAMIC_POINTS;
        else if (isSplat(segment)) score += SPLAT_POINTS;
        else score += STATIC_POINTS;
        return score;
      }, 0);
  return { route, score, index };
};

let rankRoutes = routes =>
  routes
    .map(rankRoute)
    .sort(
      (a, b) =>
        a.score < b.score
          ? 1
          : a.score > b.score ? -1 : a.index - b.index
    );

let segmentize = uri =>
  uri
    // strip starting/ending slashes
    .replace(/(^\/+|\/+$)/g, "")
    .split("/");

let addQuery = (pathname, query) =>
  pathname + (query ? `?${query}` : "");
