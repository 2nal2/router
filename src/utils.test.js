import { pick, match, resolve } from "./utils";

test("pick", () => {
  expect(pick(routes, "/").route.value).toBe("Root");
  expect(pick(routes, "/groups/main/users/me").route.value).toBe(
    "MainGroupMe"
  );
  expect(pick(routes, "/groups/123/users/456").route.value).toBe(
    "GroupUser"
  );
  expect(pick(routes, "/one/two/three/four/five").route.value).toBe(
    "Fiver"
  );
  expect(pick(routes, "/groups/main/users").route.value).toBe(
    "MainGroupUsers"
  );
  expect(pick(routes, "/groups/123/users").route.value).toBe(
    "GroupUsers"
  );
  expect(pick(routes, "/groups/main").route.value).toBe("MainGroup");
  expect(pick(routes, "/groups/123").route.value).toBe("Group");
  expect(pick(routes, "/groups").route.value).toBe("Groups");
  expect(pick(routes, "/files/some/long/path").route.value).toBe(
    "FilesDeep"
  );
  expect(pick(routes, "/files").route.value).toBe("Files");
  expect(pick(routes, "/no/where").route.value).toBe("Default");
});

test("match", () => {
  expect(match("/foo/:bar", "/foo/hello")).toMatchSnapshot();
});

test("pick return value", () => {
  expect(pick(routes, "/one/two/three/four/five")).toMatchSnapshot();
});

test("splat return value", () => {
  expect(pick(routes, "/files/some/deep/path")).toMatchSnapshot();
});

test("dynamic segments + splat return value", () => {
  let routes = [{ pattern: "/users/:userId/files/*" }];
  expect(
    pick(routes, "/users/ryan/files/some/deep/path")
  ).toMatchSnapshot();
});

test("query strings", () => {
  let routes = [{ pattern: "/users/:userId" }];
  expect(pick(routes, "/users/ryan?cool=stuff")).toMatchSnapshot();
});

test("resolve", () => {
  expect(resolve("/somewhere/else", "/users/123")).toEqual(
    "/somewhere/else"
  );
  expect(resolve("456", "/users/123")).toEqual("/users/456");
  expect(resolve("./", "/users/123")).toEqual("/users");
  expect(resolve(".", "/users/123")).toEqual("/users");
  expect(resolve("../", "/users/123")).toEqual("/");
  expect(resolve("..", "/users/123")).toEqual("/");
  expect(resolve("../.././3", "/u/1/g/4")).toEqual("/u/3");
  expect(resolve("../.././3?beef=boof", "/u/1/g/4")).toEqual(
    "/u/3?beef=boof"
  );
  expect(resolve("../.././3", "/u/1/g/4?beef=boof")).toEqual("/u/3");
  expect(resolve("stinky/barf", "/u/1/g/4")).toEqual(
    "/u/1/g/stinky/barf"
  );
  expect(resolve("?some=query", "/users/123?some=thing")).toEqual(
    "/users/123?some=query"
  );
  expect(resolve("/groups?some=query", "/users?some=thing")).toEqual(
    "/groups?some=query"
  );
});

const routes = shuffle([
  {
    value: "MainGroupMe",
    pattern: "/groups/main/users/me"
  },
  {
    value: "GroupMe",
    pattern: "/groups/:groupId/users/me"
  },
  {
    value: "GroupUser",
    pattern: "/groups/:groupId/users/:userId"
  },
  {
    value: "Fiver",
    pattern: "/:one/:two/:three/:four/:five"
  },
  {
    value: "GroupUsersSplat",
    pattern: "/groups/:groupId/users/*"
  },
  {
    value: "MainGroupUsers",
    pattern: "/groups/main/users"
  },
  {
    value: "GroupUsers",
    pattern: "/groups/:groupId/users"
  },
  {
    value: "MainGroup",
    pattern: "/groups/main"
  },
  {
    value: "Group",
    pattern: "/groups/:groupId"
  },
  {
    value: "Groups",
    pattern: "/groups"
  },
  {
    value: "FilesDeep",
    pattern: "/files/*"
  },
  {
    value: "Files",
    pattern: "/files"
  },
  {
    value: "Root",
    pattern: "/"
  },
  {
    value: "Default",
    default: true
  }
]);

function shuffle(a) {
  for (let i = a.length - 1; i > 0; i--) {
    let j = Math.floor(Math.random() * (i + 1));
    [a[i], a[j]] = [a[j], a[i]];
  }
  return a;
}
