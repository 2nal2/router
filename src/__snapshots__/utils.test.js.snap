// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dynamic segments + splat return value 1`] = `
Object {
  "params": Object {
    "*": "some/deep/path",
    "userId": "ryan",
  },
  "route": Object {
    "pattern": "/users/:userId/files/*",
  },
  "uri": "/users/ryan/files",
}
`;

exports[`match 1`] = `
Object {
  "params": Object {
    "bar": "hello",
  },
  "route": Object {
    "pattern": "/foo/:bar",
  },
  "uri": "/foo/hello",
}
`;

exports[`pick return value 1`] = `
Object {
  "params": Object {
    "five": "five",
    "four": "four",
    "one": "one",
    "three": "three",
    "two": "two",
  },
  "route": Object {
    "pattern": "/:one/:two/:three/:four/:five",
    "value": "Fiver",
  },
  "uri": "/one/two/three/four/five",
}
`;

exports[`query strings 1`] = `
Object {
  "params": Object {
    "userId": "ryan",
  },
  "route": Object {
    "pattern": "/users/:userId",
  },
  "uri": "/users/ryan",
}
`;

exports[`splat return value 1`] = `
Object {
  "params": Object {
    "*": "some/deep/path",
  },
  "route": Object {
    "pattern": "/files/*",
    "value": "FilesDeep",
  },
  "uri": "/files",
}
`;
